source/MaterialXGenShader/ShaderGraph.cpp

```diff
@@ -109,78 +200,22 @@ void ShaderGraph::addUpstreamDependencies(const Element& root, ConstMaterialPtr
-            // output itself as connecting element, which handles finding
-            // the nodedef output in case of a multioutput node upstream.
-            connectingElement = downstreamElement->asA<Output>();
-        }
-        OutputPtr nodeDefOutput = connectingElement ? upstreamNode->getNodeDefOutput(connectingElement) : nullptr;
-        ShaderOutput* output = nodeDefOutput ? newNode->getOutput(nodeDefOutput->getName()) : newNode->getOutput();
-        if (!output)
-        {
-            throw ExceptionShaderGenError("Could not find an output named '" + (nodeDefOutput ? nodeDefOutput->getName() : string("out")) +
-                "' on upstream node '" + upstreamNode->getName() + "'");
-        }
+        downstreamElements.push_back(downstreamElement);
+        upstreamElements.push_back(upstreamElement);
+        connectingElements.push_back(edge.getConnectingElement());
+    }
 
-        // First check if this was a bind input connection
-        // In this case we must have a root node as well
-        if (rootNode && connectingElement && connectingElement->isA<BindInput>())
-        {
-            // Connect to the corresponding input on the root node
-            ShaderInput* input = rootNode->getInput(connectingElement->getName());
-            if (input)
-            {
-                input->breakConnection();
-                input->makeConnection(output);
-            }
-        }
-        else
-        {
-            // Check if it was a node downstream
-            NodePtr downstreamNode = downstreamElement->asA<Node>();
-            if (downstreamNode)
-            {
-                // We have a node downstream
-                ShaderNode* downstream = getNode(downstreamNode->getName());
-                if (downstream && connectingElement)
-                {
-                    ShaderInput* input = downstream->getInput(connectingElement->getName());
-                    if (!input)
-                    {
-                        throw ExceptionShaderGenError("Could not find an input named '" + connectingElement->getName() +
-                            "' on downstream node '" + downstream->getName() + "'");
-                    }
-                    input->makeConnection(output);
-                }
-            }
-            else
-            {
-                // Not a node, then it must be an output
-                ShaderGraphOutputSocket* outputSocket = getOutputSocket(downstreamElement->getName());
-                if (outputSocket)
-                {
-                    outputSocket->makeConnection(output);
-                }
-            }
-        }
+    for (size_t i = 0; i < downstreamElements.size(); i++)
+    {
+        ElementPtr downstreamElement = downstreamElements[i];
+        ElementPtr upstreamElement = upstreamElements[i];
+        ElementPtr connectingElement = connectingElements[i];
+
+        createConnectedNodes(downstreamElement,
```

Comment:
I'm not sure I understand the purpose of first creating these three vectors and then iterating them to call createConnectedNodes. Why not call createConnectedNodes directly in the first pass?

Reply:
Agreed. It's not need. I used to have special purpose code to do the shadernode traversal and return another set of lists which where merged before calling this. I'll remove the temporary storage.